### PR TITLE
[Wayland-platform] Compositing shouldn't block in eglSwapBuffers()

### PR DESCRIPTION
**Pull request by AlanGriffiths**
[Wayland-platform] Compositing shouldn't block in eglSwapBuffers(). (Fixes: #1100)

### CODE REVIEW

**Review by mariogrip — APPROVED** [review]
LGTM :+1: 

This works really well on pinephone and pinetab :)

**Comment by AlanGriffiths** [comment]
Don't land this until we've collected some performance benchmarks for the hosted configuration.

**Comment by mariogrip** [comment]
Tested with glmark2, i have a over 100% improvements

from 200FPS (without this patch) to almost native (without nested) 600FPS

Same story on arm devices.

**Comment by AlanGriffiths** [comment]
> I _still_ don't know why this makes a difference. It looks like it _should_ just be moving from blocking in `swap_buffers` to blocking in `post`, and when I trace the callstack the `post` call gets made immediately after the call to `swap_buffers` anyway.

I was thinking that `scene_elements_for()` locked the client BufferStreams and that these are released before the `post()` call. I'll look more closely.

> Something else is happening. Notably, the X11 platform doesn't suffer from the “locked at 60FPS” problem and _it_ blocks in `eglSwapBuffers` in `swap_buffers`.

I think that is down to the X11 server accepting the buffer far more quickly than does a Mir Wayland server (which waits for the host to consume the buffer). I did try a change to that effect that (rightly) got rejected when I first discovered this issue.

**Comment by RAOF** [comment]
FWIW I've been testing against GNOME Shell, not Mir, as the host compositor.

eglSwapBuffers *should* block until vblank in both the Wayland and X11 cases.

**Comment by AlanGriffiths** [comment]
I'll get back to this discussion after I've dealt with #1700 and #1696

**Comment by AlanGriffiths** [comment]
> I _still_ don't know why this makes a difference. It looks like it _should_ just be moving from blocking in `swap_buffers` to blocking in `post`, and when I trace the callstack the `post` call gets made immediately after the call to `swap_buffers` anyway.
> 
> Something else is happening. Notably, the X11 platform doesn't suffer from the “locked at 60FPS” problem and _it_ blocks in `eglSwapBuffers` in `swap_buffers`.

My earlier understanding of why this works was definitely wrong.

I moved the `cv.wait()` from `post()` to `swap_buffers()` and still see the performance fix.

**Comment by AlanGriffiths** [comment]
@RAOF I don't understand why this works. The original code has `eglSwapBuffers()` implicitly synchronized to the frame, the new code explicitly synchronizes `swap_buffers()` to the frame.

The difference between theory and practice. Again.

**Comment by RAOF** [comment]
I'm ok with landing this, but before landing this I think we should manually check (maybe with wayland protocol debugging?) that this isn't doing something silly, like rendering unthethered to vsync.

:woman_shrugging:

**Comment by AlanGriffiths** [comment]
> I'm ok with landing this, but before landing this I think we should manually check (maybe with wayland protocol debugging?) that this isn't doing something silly, like rendering unthethered to vsync.

I *think* we're OK: I don't see anything that concerns me.

But we're basically dependent on the host compositor to handle vsync, so what does a successful check of this look like to you?

**Comment by RAOF** [comment]
Ok. I think we can land this and then see if anything weird shakes out.

bors r+

**Pull request merged**
